adapters/utils/runAdapter.ts

```diff
@@ -370,3 +376,49 @@ async function _runAdapter({
   }
 
 }
+
+function createBalanceFrom(options: {chain: string, timestamp: number | undefined, amount: FetchResponseValue}): Balances {
+  const { chain, timestamp, amount } = options
+
+  const balance = new Balances({ chain, timestamp })
+  if (amount) {
+    if (typeof amount === 'number' || typeof amount === 'string') {
+      balance.addUSDValue(amount)
+    } else {
+      balance.addBalances(amount)
+    }
+  }
+  return balance;
+}
+
+function subtractBalance(options: {balance: Balances, amount: FetchResponseValue}) {
+  const { balance, amount } = options
+  if (amount) {
+    if (typeof amount === 'number' || typeof amount === 'string') {
+      const otherBalance = createBalanceFrom({chain: balance.chain, timestamp: balance.timestamp, amount})
+      balance.subtract(otherBalance)
+    } else {
+      balance.subtract(amount)
+    }
+  }
+}
+
+function validateAdapterResult(result: any) {
+  // validate metrics
+  if (result.dailyFees) {
+    // should include atleast SupplySideRevenue or ProtocolRevenue or Revenue
+    if (!result.dailySupplySideRevenue && !result.dailyProtocolRevenue && !result.dailyRevenue) {
+      throw Error('found dailyFees record but missing all dailyRevenue, dailySupplySideRevenue, dailyProtocolRevenue records')
+    }
+  }
+}
+
+function addMissingMetrics(chain: string, result: any) {
+  if (result.dailyFees && result.dailyFees.hasBreakdownBalances()) {
```

Comment:
need to add a check here if result.dailyFees is BalancesV2 object, else the code will break when it is a number/string type

Reply:
can you also add a comment explaining why we have a `hasBreakdownBalances`, i.e this is to ensure that we dont change behavior of existing adapters